lib/Importer/pytorch/PyTorchModelLoader.cpp

```diff
@@ -0,0 +1,115 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "PyTorchModelLoader.h"
+
+namespace {
+glow::Type ptTypeToGlowType(const at::CompleteTensorType &ptType) {
+  // TODO: get correct ElemKind
+  std::vector<size_t> dims;
+  for (auto &size : ptType.sizes()) {
+    dims.push_back(size);
+  }
+  return glow::Type(glow::ElemKind::FloatTy, dims);
+}
+} // namespace
+
+glow::Placeholder *PyTorchModelLoader::loadValue(const torch::jit::Value *val) {
+  // TODO: do we need to care about optional IValues here?
+  assert(val->isCompleteTensor());
+  auto ptType = val->type()->cast<at::CompleteTensorType>();
+  auto glowType = ptTypeToGlowType(*ptType.get());
+  glow::Placeholder *ph =
+      mod_.createPlaceholder(&glowType, "input", /*isTrainable*/ false);
+  valueMap_[val] = ph->getOutput();
+  return ph;
+}
+
+void PyTorchModelLoader::loadNode(const torch::jit::Node *ptNode) {
+  glow::Node *glowNode = nullptr;
+
+  auto inputs = ptNode->inputs();
+  auto outputs = ptNode->outputs();
+
+  switch (ptNode->kind()) {
+  case at::aten::mul: {
+    assert(inputs.size() == 2);
+    glow::NodeValue rhs = valueMap_.at(inputs[0]);
+    glow::NodeValue lhs = valueMap_.at(inputs[1]);
+    glowNode = loadRes_->f->createMul("mul", rhs, lhs);
+    valueMap_[outputs[0]] = glowNode->getNthResult(glow::MulNode::ResultIdx);
+    break;
+  }
+  case at::aten::div: {
+    assert(inputs.size() == 2);
+    glow::NodeValue rhs = valueMap_.at(inputs[0]);
+    glow::NodeValue lhs = valueMap_.at(inputs[1]);
+    glowNode = loadRes_->f->createDiv("div", rhs, lhs);
+    valueMap_[outputs[0]] = glowNode->getNthResult(glow::DivNode::ResultIdx);
+    break;
+  }
+  default:
+    assert(false && "unrecognized node");
+  }
+}
+
+std::unique_ptr<ModelLoaderResult>
+PyTorchModelLoader::load(at::ArrayRef<torch::jit::IValue> &inputs) {
+  loadRes_.reset(new ModelLoaderResult());
+  static int32_t nextFuncId = 0;
+  loadRes_->f =
+      mod_.createFunction(glow::strFormat("PyTorchFunction_%d", nextFuncId++));
+
+  auto subgraphInputValues = subgraph_->inputs();
+
+  assert(inputs.size() == subgraphInputValues.size());
+
+  for (size_t i = 0; i < subgraphInputValues.size(); ++i) {
+    torch::jit::Value *inputValue = subgraphInputValues[i];
+    const c10::IValue inputIValue = inputs[i];
+    inputValue->inferTypeFrom(inputIValue.toTensor());
+    loadRes_->inputPlaceholders.push_back(loadValue(inputValue));
+  }
+
+  // Nodes are topologically sorted.
```

Comment:
Is this always guaranteed by pytorch?

Reply:
As far as I know yes, but only know this from reading other code that depends on this property